### DISCUSSION TITLE
Howto use combined condition for a mask

### DISCUSSION THREAD

**Discussion by ewelot** [libvips/libvips#2274] [Q&A] [answered]
I'd like to use a more elaborated condition to create a mask image like in the following example
```
#!/usr/bin/python3
  
import pyvips

img = pyvips.Image.xyz(3, 3)

print('image:')
for y in range(0, img.height):
    for x in range(0, img.width):
        print(x, y, img.getpoint(x,y))

mask = (img[0] > 0).ifthenelse(1,0)
print('simple mask: img[0] > 0')
for y in range(0, img.height):
    for x in range(0, img.width):
        print(x, y, mask.getpoint(x,y))

mask = (img[0] > 0 and img[0] < 2).ifthenelse(1,0)
print('more complex mask: img[0] > 0 and img[0] < 2')
for y in range(0, img.height):
    for x in range(0, img.width):
        print(x, y, mask.getpoint(x,y))

```

Unfortunately the last expression `img[0] > 0 and img[0] < 2` is not evaluated as expected but just ignores the first term. Is this intended behaviour and what is the correct way to create a combined condition (possibly beeing more complex and involving several image bands)?

**Comment by jcupitt** [comment, marked as answer]
`and` is the Python short-circuiting logical (not bitwise) AND operator, so the right-hand side is not evaluated if the if the left-hand side reduces to not `True`. For example:

```
>>> False and 1 / 0
False
```

So the `1 / 0` is never computed. 

You need to evaluate both sides to images and then find the bitwise (not logical) AND of the image pixels, eg.

```python
((img[0] > 12) & (img[1] < 99)) | ((img[1] > [1, 2, 3]) ^ (img[2] > img[1]))
```

*edit*: added brackets

**Reply by ewelot** [reply]
But isn't my example expression `img[0] > 0 and img[0] < 2` made of pure logical terms? Looking e.g. at pixel (0, 0) the program output is
```
image:
0 0 [0.0, 0.0]
...
simple mask: img[0] > 0
0 0 [0.0]
...
more complex mask: img[0] > 0 and img[0] < 2
0 0 [1.0]
...
```
My understanding is that for the given pixel the expression `img[0] > 0`becomes `0 > 0`which evaluates to `False` and therefore `img[0] > 0 and img[0] < 2` should evaluate to `False` as well. Where is my fault?

**Comment by jcupitt** [comment]
In pyvips, `image > 0` returns a boolean image with 0 for all false pixels and `255` for all true pixels.

Do you want to detect if every pixel in the image is `> 0`? You'd need something like `(img > 0).min() != 0`, ie. the minimum pixel value of the relation is not zero.

**Reply by jcupitt** [reply]
You can see the behaviour at the py prompt:

```
$ python3
Python 3.8.6 (default, Jan 27 2021, 15:42:20) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("pics/k2.jpg")
>>> x > 128
<pyvips.Image 1450x2048 uchar, 3 bands, srgb>
>>> x(0, 0)
[11.0, 1.0, 0.0]
>>> (x > 10)(0, 0)
[255.0, 0.0, 0.0]
>>> (x > [20, 0, 0])(0, 0)
[0.0, 255.0, 0.0]
>>> 
```

**Reply by ewelot** [reply]
Thanks for your answer. I believe I do understand how a single condition on pixel values works. What I would like to create in my example program is somewhat different. I'd like to get a "boolean" mask image which has 1 (meaning true) for all pixels where the value in the first band of the input image is both greater than 0 **and** less than 2. For any other pixel the mask image should contain a value of 0 (meaning false). Sorry, this was probably not very clear in my initial post.

May I ask again for a suggestion on how to accomplish this in pyvips?
Thanks for help.

**Reply by ewelot** [reply]
Maybe one possible solution is to use `bandrank` to get the lowest value for each pixel with an array of images, each describing one constraint  ...

EDIT: Adding brackets to the example provided by the first answer did the trick! I'll tag it as best solution now ...

**Reply by jcupitt** [reply]
Yes, that would just be:

```python
(img[0] > 0) & (img[0] < 2)
```

ie. all pixels with the value 1, I suppose.

*edit* added brackets

**Reply by ewelot** [reply]
That's it! Adding the brackets was the clue ...
Thanks again!

How do I tag your last reply as the right answer?

**Comment by jcupitt** [comment]
Perhaps this helps? You can write something to count the number of true pixels like this:

```python
def count_set(img):
     return img.avg() * img.width * img.height * img.bands / 255.0
```

Then:

```
>>> count_set(x[0] == 1)
9858.0
>>> count_set((x[0] > 0) & (x[0] < 2))
9858.0
```
